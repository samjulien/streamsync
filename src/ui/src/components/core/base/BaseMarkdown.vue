<template>
	<template v-if="html !== undefined">
		<span
			v-if="inline"
			v-dompurify-html="html"
			class="BaseMarkdown markdown"
		></span>
		<div v-else v-dompurify-html="html" class="BaseMarkdown markdown"></div>
	</template>
</template>

<script setup lang="ts">
import { ref, toRefs, watch } from "vue";

const props = defineProps({
	rawText: { type: String, required: true },
	inline: { type: Boolean },
});

const html = ref<string | undefined>(undefined);

const { rawText, inline } = toRefs(props);

async function buildHtml() {
	const marked = await import("marked");
	const result = inline.value
		? marked.parseInline(rawText.value, { async: false })
		: marked.parse(rawText.value, { async: false });
	html.value = await result;
}

watch([rawText, inline], buildHtml, { immediate: true });
</script>

<style scoped>
.markdown {
	font-size: 0.875rem;
}

.markdown:deep() h1,
.markdown:deep() h2,
.markdown:deep() h3,
.markdown:deep() h4 {
	margin: 0;
	color: var(--primaryTextColor);
}

.markdown:deep() h1 {
	font-size: 1.5rem;
	letter-spacing: 0.1875rem;
	font-weight: 600;
	line-height: 162.5%;
}

.markdown:deep() h2 {
	font-size: 1.5rem;
	font-weight: 500;
	line-height: 120%;
}

.markdown:deep() h3 {
	font-size: 1.25rem;
	font-weight: 500;
	line-height: 120%;
}

.markdown:deep() h4 {
	font-weight: 700;
	font-size: 1rem;
	line-height: 247.5%;
}

.markdown:deep() ul,
.markdown:deep() ol {
	padding: 0;
	padding-inline-start: 0;
	margin-block-start: 0;
}

.markdown:deep() li {
	margin: 0 0 0 32px;
}

.markdown:deep() hr {
	border: none;
	border-top: 1px solid var(--separatorColor);
}

.markdown:deep() pre {
	background-color: var(--separatorColor);
	font-family: monospace;
	padding: 8px;
}

.markdown:deep() code {
	background-color: var(--separatorColor);
	font-family: monospace;
	padding: 2px;
}

.markdown:deep() pre > code {
	background-color: unset;
}

.markdown:deep() table {
	border-collapse: collapse;
}

.markdown:deep() th {
	padding: 8px;
	background: var(--separatorColor);
	border: 1px solid var(--separatorColor);
}

.markdown:deep() td {
	padding: 8px;
	border: 1px solid var(--separatorColor);
}
</style>
