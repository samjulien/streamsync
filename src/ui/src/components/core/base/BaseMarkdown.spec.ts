import { beforeEach, describe, expect, it, Mock, vi } from "vitest";
import { flushPromises, shallowMount } from "@vue/test-utils";
import VueDOMPurifyHTML from "vue-dompurify-html";
import BaseMarkdown from "./BaseMarkdown.vue";
import { parse, parseInline } from "marked";

vi.mock("marked", () => ({
	parse: vi.fn().mockImplementation((v) => `<div>${v}</div>`),
	parseInline: vi.fn().mockImplementation((v) => `<span>${v}</span>`),
}));

describe("BaseMarkdown", () => {
	const rawText = "# hello\nI'm **MD** ";

	beforeEach(() => {
		(parse as unknown as Mock).mockClear();
		(parseInline as unknown as Mock).mockClear();
	});

	it("should render markdown as block", async () => {
		const wrapper = shallowMount(BaseMarkdown, {
			props: { rawText },
			global: { plugins: [VueDOMPurifyHTML] },
		});
		await flushPromises();
		expect(wrapper.element).toMatchSnapshot();
		expect(parse).toHaveBeenCalledOnce();

		wrapper.setProps({ rawText: "updated" });
		await flushPromises();
		expect(parse).toHaveBeenCalledTimes(2);
	});

	it("should render markdown as inline", async () => {
		const wrapper = shallowMount(BaseMarkdown, {
			props: { rawText, inline: true },
			global: { plugins: [VueDOMPurifyHTML] },
		});
		await flushPromises();
		expect(wrapper.element).toMatchSnapshot();
		expect(parseInline).toHaveBeenCalledOnce();

		wrapper.setProps({ rawText: "updated" });
		await flushPromises();
		expect(parseInline).toHaveBeenCalledTimes(2);
	});
});
