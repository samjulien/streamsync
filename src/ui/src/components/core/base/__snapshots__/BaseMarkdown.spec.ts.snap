// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseMarkdown > should render markdown as block 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="BaseMarkdown markdown"
    data-v-03a7c885=""
  >
    <div>
      # hello
I'm **MD** 
    </div>
  </div>
  
</div>
`;

exports[`BaseMarkdown > should render markdown as inline 1`] = `
<div
  data-v-app=""
>
  
  <span
    class="BaseMarkdown markdown"
    data-v-03a7c885=""
  >
    <span>
      # hello
I'm **MD** 
    </span>
  </span>
  
</div>
`;
